### PR TITLE
oid->GlobalId in ScalarType

### PR DESCRIPTION
**Pull request by sploiselle**
The bulk of #11344 is using GlobalIds in ScalarType; this PR does that. Note that we should not close the linked-to issue because there is still work to unit the concept of a custom name and a custom OID for `ScalarType::Record`.

### Motivation

This PR refactors existing code.

### Tips for reviewer

The vast majority of this PR is code movement.

This does undo some of @jkosh44 's work from #11643––couldn't quite understand that code's motivation in light of these changes, so removed it, though it's possible it teed up other work that will now be impeded by its absence? Let me know.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - there are no user-facing behavior changes

### CODE REVIEW

File: src/coord/src/catalog/builtin.rs
Comment (jkosh44) [inline src/coord/src/catalog/builtin.rs:132]: We should probably add back in this comment below
https://github.com/MaterializeInc/materialize/blob/3c9ea8fb70154c2ef72addddcb864e38a0d74dc8/src/coord/src/catalog/builtin.rs#L135-L139

Comment (sploiselle) [reply src/coord/src/catalog/builtin.rs:132]: Good call; will look at these docs

File: src/sql/src/catalog.rs
Comment (jkosh44) [inline src/sql/src/catalog.rs:383]: This used to be called `element_id`, might make sense to rename it back.

File: src/sql/src/catalog.rs
Comment (jkosh44) [inline src/sql/src/catalog.rs:397]: This used to be called `element_id`, might make sense to rename it back.

File: src/sql/src/catalog.rs
Comment (jkosh44) [inline src/sql/src/catalog.rs:401]: These used to be called `key_id` and `value_id`, might make sense to rename them back.

File: src/sql/src/func.rs
Comment (jkosh44) [inline src/sql/src/func.rs:439]: Should these all be names `X_ids` instead of `X_oids` or am I misunderstanding?

**Review by benesch — COMMENTED** [review]
I would very much like to get us back to dynamically allocating type IDs eventually, but I'm comfortable merging this as is if @jkosh44 is!

File: src/coord/src/catalog/storage.rs
Comment (jkosh44) [inline src/coord/src/catalog/storage.rs:28]: In the `MIGRATIONS` sql statements near line 60 I think we should remove all types from the table `system_gid_mapping`

**Comment by jkosh44** [comment]
@sploiselle given what @benesch said, would you be able to expand on what part of the dynamic IDs were making this difficult? If you have the type name, schema, and a reference to the catalog then you should be able to resolve the type global ID fairly easily. My guess is that a catalog reference is not easily available when you need to get the ID?

File: src/coord/src/catalog/builtin.rs
Comment (sploiselle) [reply src/coord/src/catalog/builtin.rs:1160]: Yeah, this was just an over-eager find/replace.

File: src/sql/src/func.rs
Comment (sploiselle) [reply src/sql/src/func.rs:439]: ty--oversight on my end

Comment (sploiselle) [reply src/sql/src/func.rs:439]: Yeah, just a missed rename

**Comment by benesch** [comment]
> @sploiselle given what @benesch said, would you be able to expand on what part of the dynamic IDs were making this difficult? If you have the type name, schema, and a reference to the catalog then you should be able to resolve the type global ID fairly easily. My guess is that a catalog reference is not easily available when you need to get the ID?

Yeah, I think there are a few moments in function selection where you want to take e.g. a `ParamType` to a global ID, and that'll require a catalog.

@jkosh44 I'm going to defer the rest of this review to you! Everything checks out for me on a skim though.

**Comment by benesch** [comment]
Thanks very much for crunching through this one, @sploiselle! It just occurred to me that this went in a little ... too ... easy, and I think we've got a bug: https://github.com/MaterializeInc/materialize/issues/11771. Not something that needs to be fixed now, I don't think! But wanted to call it out while you have this context and get your sense for whether that issue description seems right.

**Review by jkosh44 — APPROVED** [review]
LGTM, I think once you merge this I'll take a look and see how we can add dynamic type IDs back in.

**Comment by sploiselle** [comment]
@jkosh44 I'm refactoring to use dynamic IDs, no sweat.

**Comment by jkosh44** [comment]
> @jkosh44 I'm refactoring to use dynamic IDs, no sweat.

Ok, in that case I'll re-review once that's done. 

If you already have a plan then feel free to ignore this. My initial thoughts are that it might be possible to add a `GlobalId` field to `ParamType`. Then during bootstrap of functions, when we initialize the `ParamType`s we can initialize them with `GlobalId`s by resolving the type ids. That's a bit hand wavy and I have no idea if it will actually work though.

**Comment by sploiselle** [comment]
@jkosh44 @mjibson This is ready for review again.

@mjibson pinging you only because I had to do some nonsense with the catalog to handle resolving types from their names only. There are other, more elaborate approaches that don't require cheating through `get_entry_in_system_schemas` (you'll see), but figured this was defensible as a private method that saved a bunch of annoying code to the same end elsewhere in `pack_func_update`.

File: src/pgrepr/src/types.rs
Comment (jkosh44) [inline src/pgrepr/src/types.rs:575]: Would it be possible to change this to 
```rust
    pub fn catalog_name(&self) -> (&'static str, &'static str) {
        (self.inner().name(), self.inner().schema())
    }
```

That way when we go to look it up in the catalog, we can check the specific system schema and not have to look in all of them.

Does the `schema()` return accurate information? I see that it's part of an external library.

File: src/coord/src/catalog.rs
Comment (jkosh44) [inline src/coord/src/catalog.rs:273]: If we did have the schema name here (see comment on `mz_pgrepr::types::Type::catalog_name`), then we wouldn't have to search all system schemas here.

**Review by jkosh44 — CHANGES_REQUESTED** [review]

**Comment by jkosh44** [comment]
@sploiselle I did not mean to submit that review as "requested changes", I was just trying to submit comments since I'm not even sure if my suggestion is possible.

File: src/repr/src/proto/global_id.rs
Comment (mjibson) [inline src/repr/src/proto/global_id.rs:15]: Doesn't this already exist? https://dev.materialize.com/api/rust-private/mz_expr/proto/id/struct.ProtoGlobalId.html

File: src/repr/src/proto/relation_and_scalar.proto
Comment (jkosh44) [reply src/repr/src/proto/relation_and_scalar.proto:44]: I think @benesch mentioned in slack somewhere yesterday that we can start breaking proto backwards compatibility until platform is stable.

File: src/repr/src/proto/global_id.rs
Comment (sploiselle) [reply src/repr/src/proto/global_id.rs:15]: https://github.com/MaterializeInc/materialize/pull/11627/files#diff-ae96a1abfa287351061df1b13da8b332caba6157b5bf30c940f71df9fcf1cdafL21

File: src/pgrepr/src/types.rs
Comment (sploiselle) [reply src/pgrepr/src/types.rs:575]: That would work if we change the crates' dependency graphs and generated `ParamType` from builtin types (not a bad idea, just work that we could elect to do). In the current world `ParamType` isn't attached to all of the builtin types, so the pseudotype parameters don't have any notion of their schema.

SQL planning doesn't currently keep such a tight hold on e.g. system schemas, so I was trying to avoid muddying that relationship any more than I needed to.

This all being said, I think we're due for a refactor that brings builtin types into something that can be more easily shared between SQL and coord, for exactly this kind of inelegance.

File: src/repr/src/proto/relation_and_scalar.proto
Comment (sploiselle) [reply src/repr/src/proto/relation_and_scalar.proto:44]: This was an unintentional break on my end--I spent a few hours wrestling with the protos and simply forgot to change it back. No harm in doing the right thing, so will fix this up.

File: src/repr/src/proto/global_id.rs
Comment (mjibson) [reply src/repr/src/proto/global_id.rs:15]: Ah, didn't see that.

File: src/pgrepr/src/types.rs
Comment (jkosh44) [reply src/pgrepr/src/types.rs:575]: Well that sounds like a lot of work for probably not a ton of benefit as far as this specific PR's functionality is concerned. So unless you're really eager to do that, it's probably best left to a later issue/PR.

**Review by jkosh44 — APPROVED** [review]

**Comment by jkosh44** [comment]
@sploiselle btw, I was seeing a similar persistence error on another PR. When I rebased over main it started passing again. So it's possible that it's not related to your PR.

**Pull request merged**
